sound/soc/sof/intel/hda-loader.c

```diff
@@ -167,9 +167,6 @@ static int cl_trigger(struct snd_sof_dev *sdev,
 	/* code loader is special case that reuses stream ops */
 	switch (cmd) {
 	case SNDRV_PCM_TRIGGER_START:
-		wait_event_timeout(sdev->waitq, !sdev->code_loading,
-				   HDA_DSP_CL_TRIGGER_TIMEOUT);
```

Comment:
If this is correct, I'd be all for this - removing redundant synchronisation controls simplifies the code. But like others I cannot be completely certain that this is safe on all platforms. This assumes, that by the time we arrive here code_loading is already cleared and we don't wait anyway. So, can we add wait with this commit and add a
```
WARN_ON(sdev->code_loading);
```
here and run with that code for some time? After we've verified that that warning never triggers, we should be more confident, that it's safe to remove the wait queue. When removing it we still can keep the WARN().

Reply:
We have 2 threads in the boot process in parallel, one is the boot thread, the other is the irq thread. once interrupt is initialized in boot thread,  "PURGE FW IPC" initiated by previous purge FW request will propagate to irq thread. This IPC happens really early, even before interrupt initialization. but we have to deal with it(call hda_dsp_ipc_dsp_done to clear some bit), or further ipc will not work. Some later polls in cl_dsp_init(hda_loader.c) cause the execution of irq thread ahead of TRIGGER START, that's the reason fw still boot after removing boot_loading. Actually, we are depending on linux scheduler, and what I am really afraid is that if the irq thread not get executed before TRIGGER START, FW surely boot failed!!!(didn't see this till now) but boot_loading guaranteed the execution order.